src/net.cpp

```diff
@@ -193,30 +193,30 @@ bool IsPeerAddrLocalGood(CNode *pnode)
            IsReachable(addrLocal.GetNetwork());
 }
 
-// pushes our own address to a peer
 void AdvertiseLocal(CNode *pnode)
 {
-    if (fListen && pnode->fSuccessfullyConnected)
+    assert(fListen);
+    // First consider our "best" addr for the peer, as seen locally.
+    CAddress addrLocal = GetLocalAddress(&pnode->addr, pnode->GetLocalServices());
+
+    if (gArgs.GetBoolArg("-addrmantest", false)) {
+        // use IPv4 loopback during addrmantest
+        addrLocal = CAddress(CService(LookupNumeric("127.0.0.1", GetListenPort())), pnode->GetLocalServices());
+    }
+
+    // If discovery is enabled, sometimes give our peer the address it
+    // tells us that it sees us as in case it has a better idea of our
+    // address than we do.
```

Comment:
This comment confuses me. If Peer_A gave us "the address it tells us that it sees us as", why would we need to give Peer_A our address?

Is this comment supposed to mean that we sometimes refer to ourselves as the address that another peer gave us? I.e. we heard from Peer_A that they know us as `addrRemote`, and we sometimes announce our address to another peer `Peer_B` as `addrRemote` instead of `addrLocal`?

Reply:
Hopefully [this](https://github.com/bitcoin/bitcoin/pull/19843#discussion_r500055283) comment answers.

Why we give it to them? Because we ask them to *forward* our self-announcement to other nodes. Peer_A is not supposed currently to modify an `ADDR` they receive, so they just take what we give them and forward it.

You are right that they could, but it has other implications...

Reply:
Okay, so we are giving `Peer_A` the address they gave us to forward to its peers. Now that makes more sense to me. :)